vpr/src/route/connection_router.cpp

```diff
@@ -330,41 +327,33 @@ vtr::vector<RRNodeId, t_heap> ConnectionRouter<Heap>::timing_driven_find_all_sho
         // lookahead we can re-use the node exploration code from the regular router
         RRNodeId target_node = RRNodeId::INVALID();
 
-        timing_driven_expand_cheapest(cheapest,
+        timing_driven_expand_cheapest(inode,
+                                      new_total_cost,
                                       target_node,
                                       cost_params,
                                       bounding_box,
                                       t_bb());
 
-        if (cheapest_paths[inode].index == RRNodeId::INVALID() || cheapest_paths[inode].cost >= cheapest->cost) {
-            VTR_LOGV_DEBUG(router_debug_, "  Better cost to node %d: %g (was %g)\n", inode, cheapest->cost, cheapest_paths[inode].cost);
-            cheapest_paths[inode] = *cheapest;
+        if (cheapest_paths[inode].index == RRNodeId::INVALID() || cheapest_paths[inode].total_cost >= new_total_cost) {
+            VTR_LOGV_DEBUG(router_debug_, "  Better cost to node %d: %g (was %g)\n", inode, new_total_cost, cheapest_paths[inode].total_cost);
+            // Only the `index` and `prev_edge` fields of `cheapest_paths[inode]` are used after this function returns
+            cheapest_paths[inode].index = inode;
+            cheapest_paths[inode].prev_edge = rr_node_route_inf_[inode].prev_edge;
         } else {
-            VTR_LOGV_DEBUG(router_debug_, "  Worse cost to node %d: %g (better %g)\n", inode, cheapest->cost, cheapest_paths[inode].cost);
+            VTR_LOGV_DEBUG(router_debug_, "  Worse cost to node %d: %g (better %g)\n", inode, new_total_cost, cheapest_paths[inode].total_cost);
         }
-
-        rcv_path_manager.free_path_struct(cheapest->path_data);
-        heap_.free(cheapest);
     }
 
     return cheapest_paths;
 }
 
 template<typename Heap>
-void ConnectionRouter<Heap>::timing_driven_expand_cheapest(t_heap* cheapest,
+void ConnectionRouter<Heap>::timing_driven_expand_cheapest(RRNodeId from_node,
+                                                           float new_total_cost,
                                                            RRNodeId target_node,
                                                            const t_conn_cost_params& cost_params,
                                                            const t_bb& bounding_box,
                                                            const t_bb& target_bb) {
-    RRNodeId inode = cheapest->index;
-
-    t_rr_node_route_inf* route_inf = &rr_node_route_inf_[inode];
-    float best_total_cost = route_inf->path_cost;
-    float best_back_cost = route_inf->backward_path_cost;
-
-    float new_total_cost = cheapest->cost;
-    float new_back_cost = cheapest->backward_path_cost;
-
     /* I only re-expand a node if both the "known" backward cost is lower  *
```

Comment:
I think this comment is not out of date (superceded by the comment below).  I think you should merge the two comments (mention that the condition below ensures the node will not be reexplored if we have found an inferior path to it as that could create electrically illegal loops, and would also degrade QoR even when it didn't form a loop).

Reply:
I see why illegal loops or QoR degradation happened in the previous algorithm (below), since we could not use an inferior path stored in the heap node to re-explore this node.
```
while pop {
   if not prune {
      update_cheapest;
      foreach edge {
          if not prune {
             push (index, costs, ...) to heap;
          }
    }
}
```

But I am not quite sure why it would happen in our current algorithm, since even if we re-explore the node, its node info is stored in `rr_node_route_info` and always represents the best path. Please correct me if I’m wrong, I guess re-exploration might just be redundant but harmless work?

Reply:
RCV often isn't searching for a shortest path; it is searching for a path in the target delay range.  So it might find a path to node n that has a higher backward_path_cost but the total_cost (including expected delay to sink, going through a cost function that checks that against the target delay) might be lower than the previously stored value.  In that case we want to re-expand the node so long as it doesn't create a loop. That rcv_path_manager should store enough info for us to avoid loops.

Reply:
Thanks for this detailed explanation! I will add extra comments for this.